ARROW-10197: [python][Gandiva] Execute expression on filtered data

KirillLykov opened the pull request:
This PR is to enable running queries on top of filtered data in python. For that, I extended Cython wrapper and added a unit test which is the same as a test in cpp module https://github.com/apache/arrow/blob/c5fa23ea0e15abe47b35524fa6a79c7b8c160fa0/cpp/src/gandiva/tests/filter_project_test.cc#L208

KirillLykov commented:
Lint failure is clear but two other failing checks:
* Python / AMD64 MacOS 10.15 Python 3
* continuous-integration/appveyor/pr

Do not seem to be related with my changes.

wesm commented:
The community is in the middle of making a release so it may be a little while before a reviewer can get to this

suquark reviewed (approved):
LGTM. The CI errors seem unrelated.

wesm commented:
@pitrou or @bkietz could one of you have a look?

bkietz commented on python/pyarrow/includes/libgandiva.pxd:
```diff
@@ -58,6 +67,31 @@ cdef extern from "gandiva/selection_vector.h" namespace "gandiva" nogil:
+    else:
+        raise ValueError('Invalid value for Selection Mode: {!r}'.format(name))
+
+cdef inline str _selection_mode_name(CSelectionVector_Mode ctype):
```
```suggestion
cdef inline str _selection_mode_name(CSelectionVector_Mode ctype) except *:
```

KirillLykov replied:
It cannot be added because `Exception clause not allowed for function returning Python object`

bkietz commented on python/pyarrow/gandiva.pyx:
```diff
@@ -97,6 +100,21 @@ cdef class Expression(_Weakrefable):
     cdef void init(self, shared_ptr[CExpression] expression):
         self.expression = expression
 
+cdef class Configuration(_Weakrefable):
+    cdef:
+        shared_ptr[CConfiguration] config
+
+    def __init__(self):
+        raise TypeError("Do not call {}'s constructor directly, use the "
+                        "TreeExprBuilder API directly"
+                        .format(self.__class__.__name__))
+
+    @staticmethod
+    cdef create():
+        cdef Configuration self = Configuration.__new__(Configuration)
+        self.config = CConfigurationBuilder.DefaultConfiguration()
+        return self
+
```
```suggestion
```

This class is currently unnecessary since it only wraps the default configuration. Exposing it can wait until there's a need for python to adjust a Configuration.

bkietz commented on python/pyarrow/gandiva.pyx:
```diff
@@ -405,9 +437,24 @@ cpdef make_projector(Schema schema, children, MemoryPool pool):
+    for child in children:
+        c_children.push_back(child.expression)
+    cdef shared_ptr[CProjector] result
+    cdef Configuration configuration = Configuration.create()
+    cdef CSelectionVector_Mode mode = _ensure_selection_mode(selection_mode)
+    check_status(
+        Projector_Make(schema.sp_schema, c_children, mode,
+                       configuration.config, &result))
```
```suggestion
    check_status(
        Projector_Make(schema.sp_schema,
                       c_children,
                       _ensure_selection_mode(selection_mode),
                       CConfigurationBuilder.DefaultConfiguration(),
                       &result))
```

bkietz commented on python/pyarrow/gandiva.pyx:
```diff
@@ -405,9 +437,24 @@ cpdef make_projector(Schema schema, children, MemoryPool pool):
                                 &result))
     return Projector.create(result, pool)
 
+cpdef make_projector_with_mode(Schema schema, children,
+                               str selection_mode, MemoryPool pool):
```
Some duplicated code could be removed by making selection_mode an optional argument of make_projector instead of writing a new function

bkietz commented on python/pyarrow/gandiva.pyx:
```diff
@@ -161,6 +179,20 @@ cdef class Projector(_Weakrefable):
             arrays.append(pyarrow_wrap_array(result))
         return arrays
 
+    def evaluate_with_selection(self, RecordBatch batch,
+                                SelectionVector selection):
```
Likewise `selection` could be an optional argument of evaluate()

KirillLykov commented:
It looks like I addressed all the comments by @bkietz . The only thing is that the commit history looks a bit too long and I would recommend squashing (don't know project policy about merging/rebasing/squashing).
